CIScripts/BuildFunctions.ps1

```diff
@@ -0,0 +1,151 @@
+
+        Write-Host $("Cloning " +  $_.Url + " from branch: " + $CustomMultiBranch)
+        git clone -b $CustomMultiBranch $_.Url $_.Dir
+
+        if ($LASTEXITCODE -ne 0) {
+            Write-Host $("Cloning " +  $_.Url + " from branch: " + $_.Branch)
+            git clone -b $_.Branch $_.Url $_.Dir
+
+            if ($LASTEXITCODE -ne 0) {
+                throw "Cloning from " + $_.Url + " failed"
+            }
+        }
+    })
+}
+
+function Contrail-Common-Actions {
+    Param ([Parameter(Mandatory = $true)] [string] $ThirdPartyCache,
+           [Parameter(Mandatory = $true)] [string] $VSSetupEnvScriptPath)
+    
+    Write-Host "Sourcing VS environment variables"
+    Invoke-BatchFile "$VSSetupEnvScriptPath"
+    
+    Write-Host "Copying common third-party dependencies"
+    New-Item -ItemType Directory .\third_party
+    Copy-Item -Recurse "$ThirdPartyCache\common\*" third_party\
+
+    Copy-Item tools\build\SConstruct .\
+}
+
+function Sign-MSI {
+    Param ([Parameter(Mandatory = $true)] [string] $SigntoolPath,
+           [Parameter(Mandatory = $true)] [string] $CertPath,
+           [Parameter(Mandatory = $true)] [string] $CertPasswordFilePath,
+           [Parameter(Mandatory = $true)] [string] $MSIPath)
+    
+    $cerp = Get-Content $CertPasswordFilePath
+    & $SigntoolPath sign /f $CertPath /p $cerp $MSIPath
+    if ($LASTEXITCODE -ne 0) {
+    throw "Signing $MSIPath failed"
+}
+}
+
+function Build-DockerDriver {
+    Param ([Parameter(Mandatory = $true)] [string] $DriverSrcPath,
+           [Parameter(Mandatory = $true)] [string] $SigntoolPath,
+           [Parameter(Mandatory = $true)] [string] $CertPath,
+           [Parameter(Mandatory = $true)] [string] $CertPasswordFilePath)
+    
+    $Env:GOPATH=pwd
+    
+    New-Item -ItemType Directory ./bin
+    Push-Location bin
+
+    Write-Host "Installing test runner"
+    go get -u -v github.com/onsi/ginkgo/ginkgo
+
+    Write-Host "Building driver"
+    go build -v $DriverSrcPath
+
+    $srcPath = "$Env:GOPATH/src/$DriverSrcPath"
+    Write-Host $srcPath
+
+    Write-Host "Precompiling tests"
+    $modules = @("driver", "controller", "hns", "hnsManager")
+    $modules.ForEach({
+        .\ginkgo.exe build $srcPath/$_
+        Move-Item $srcPath/$_/$_.test ./
+    })
+
+    Write-Host "Copying Agent API python script"
+    Copy-Item $srcPath/scripts/agent_api.py ./
+
+    Write-Host "Intalling MSI builder"
+    go get -u -v github.com/mh-cbon/go-msi
+
+    Write-Host "Building MSI"
+    Push-Location $srcPath
+    & "$Env:GOPATH/bin/go-msi" make --msi installer.msi --arch x64 --version 0.1 --src template --out $pwd/gomsi
+    Pop-Location
+
+    Move-Item $srcPath/installer.msi ./
+    
+    Sign-MSI -SigntoolPath $SigntoolPath -CertPath $CertPath -CertPasswordFilePath $CertPasswordFilePath -MSIPath "installer.msi"
+
+    Pop-Location
+}
+
+function Build-Extension {
```

Comment:
Please use approved words in function names (https://msdn.microsoft.com/en-us/library/ms714428(v=vs.85).aspx)